biojava-structure/src/main/java/org/biojava/nbio/structure/basepairs/BasePairParameters.java

```diff
@@ -0,0 +1,576 @@
+     * @param removeDups whether to only look for base-pair parameters for each unique sequence in
+     *  the structure (if set to <i>true</i>)
+     * @param canonical Whether to consider only Watson-Crick base pairs
+     */
+    public BasePairParameters(Structure structure, boolean useRNA, boolean removeDups, boolean canonical) {
+        this.structure = structure;
+        this.useRNA = useRNA;
+        this.canonical = canonical;
+        this.nonredundant = removeDups;
+
+    }
+
+    public BasePairParameters(Structure structure, boolean useRNA, boolean removeDups) {
+        this(structure, useRNA, removeDups, false);
+    }
+
+    public BasePairParameters(Structure structure, boolean useRNA) {
+        this(structure, useRNA, false, false);
+    }
+
+    /**
+     * Constructor takes a Structure object, finds base pair and base-pair step parameters
+     * for double-helical regions within the structure for only canonical DNA pairs.
+     * @param structure The already-loaded structure to analyze.
+     */
+    public BasePairParameters(Structure structure) {
+        this(structure, false, false, true);
+    }
+
+
+    /**
+     * This is the main function call to extract all step parameters, pairing parameters, and sequence
+     * information from the Structure object provided to the constructor.
+     * @return This same object with the populated data, convenient for output
+     *  (e.g. <i>log.info(new BasePairParameters(structure).analyze());</i>)
+     */
+    public BasePairParameters analyze() {
+        if (structure == null) {
+            pairingParameters = null;
+            stepParameters = null;
+            return this;
+        }
+        List<Chain> nucleics = this.getNucleicChains(nonredundant);
+        List<Group[]> pairs = this.findPairs(nucleics);
+        this.pairingParameters = new double[pairs.size()][6];
+        this.stepParameters = new double[pairs.size()][6];
+        Matrix4d lastStep;
+        Matrix4d currentStep = null;
+        for (int i = 0; i < pairs.size(); i++) {
+            lastStep = currentStep;
+            currentStep = this.basePairReferenceFrame(pairs.get(i));
+            referenceFrames.add((Matrix4d)currentStep.clone());
+            for (int j = 0; j < 6; j++) pairingParameters[i][j] = pairParameters[j];
+            if (i != 0) {
+                lastStep.invert();
+                lastStep.mul(currentStep);
+                double[] sparms = calculatetp(lastStep);
+                for (int j = 0; j < 6; j++) stepParameters[i][j] = sparms[j];
+            }
+        }
+        return this;
+    }
+
+
+    /**
+     * This reports all the pair parameters, in the order of:
+     * buckle, propeller, opening (in degrees), shear, stagger, stretch (in Å).
+     * @return A double[][] with length equal to number of base pairs for rows, and 6 columns
+     */
+    public double[][] getPairingParameters() {
```

Comment:
In general, I would not recommend returning nested arrays directly, but rather create getter methods for each of the parameter arrays. This would be a much friendlier API for the user and it can avoid trivial bugs and nested loops to change the data. For example, we could have: `getBuckle(), getPropeller(), getOpening(),` etc.

You might keep the internal data structures and this getter method, but add additional getter methods for each.